File: 24Dot.py

numbers=[4, 9, 10, 12]

def Search(numbers):
    expression = list(map(str, numbers))
    def _Search(n):
        if n==0:
            return numbers[n] == 24

        for i in range(n+1):
            for j in range(i+1,n+1):
                a, b = numbers[i],numbers[j]
                expa, expb = expression[i],expression[j]

                numbers[j] = numbers[n]
                expression[j] = expression[n]

                expression[i] = '(%s+%s)' % (expa,expb)
                numbers[i] = a + b
                if _Search(n - 1):
                    return True

                expression[i] = '(%s-%s)' % (expa,expb)
                numbers[i] = a - b
                if _Search(n - 1):
                    return True

                expression[i] = '(%s-%s)' % (expb,expa)
                numbers[i] = b - a
                if _Search(n - 1):
                    return True

                
                expression[i] = '(%s*%s)' % (expa,expb)
                numbers[i] = a * b
                if _Search(n - 1):
                    return True

                if b != 0:
                    expression[i] = '(%s/%s)' % (expa,expb)
                    numbers[i] = a / b
                    if _Search(n - 1):
                        return True
                    
                if a != 0:
                    expression[i] = '(%s*%s)' % (expb,expa)
                    numbers[i] = b / a
                    if _Search(n - 1):
                        return True

                numbers[i],numbers[j] = a,b
                expression[i],expression[j] = expa,expb
        return False

    return expression[0] if _Search(len(numbers) -1) else "FAIL"

def main():
    print(Search(numbers))

if __name__ == '__main__':
    main()

File: 24dot.lua
local number={
4,
9 ,
10 , 
12
}
local expression={}

local PRECISION = 0.000001
local function toExpression(numbers)
	if not numbers or type(numbers) ~= 'table' then return end
	for i,v in ipairs(numbers) do 
		expression[i]= tostring(v);
	end
end

local function Search(n) 
    if (n == 1) then 
    	return math.abs(number[1] - 24)<PRECISION
    end

    for i=1,n do 
        for j=i+1,n do
            local a, b = number[i],number[j]
            local expa, expb = expression[i],expression[j]

            number[j] = number[n]
            expression[j] = expression[n]

            expression[i] = '(' .. expa .. '+' .. expb .. ')'; 
            number[i] = a + b; 
            if Search(n - 1)  then return true end

            expression[i] = '(' .. expa .. '-' .. expb .. ')'
            number[i] = a - b; 
            if Search(n - 1)  then return true end

            expression[i] = '(' .. expb .. '-' .. expa .. ')'
            number[i] = b - a; 
            if Search(n - 1)  then return true end


            expression[i] = '(' .. expa .. '*' .. expb .. ')' 
            number[i] = a * b; 
            if Search(n - 1)  then return true end

            if (b ~= 0) then 
                expression[i] = '(' .. expa .. '/' .. expb .. ')' 
                number[i] = a / b; 
                if Search(n - 1)  then return true end
            end
            if (a ~= 0) then
                expression[i] = '(' .. expb .. '/' .. expa .. ')' 
                number[i] = b / a; 
                if Search(n - 1)  then return true end
            end

            number[i] = a
            number[j] = b
            expression[i] = expa
            expression[j] = expb
        end
    end
    return false; 
end

local function main()
	toExpression(number)
	print(Search(#number) and "SUCCESS\n" .. expression[1] or "FAIL")
end
main()